test/upgrade/prober/verify.go

```diff
@@ -84,6 +108,45 @@ func (p *prober) Finish() {
 	p.removeSender()
 }
 
+func (p *prober) getTraceForStepEvent(eventNo string) []byte {
+	query := fmt.Sprintf("step=%s and cloudevents.type=%s and target=%s",
+		eventNo, event.StepType, fmt.Sprintf(forwarderTargetFmt, p.client.Namespace))
+	trace, err := event.FindTrace(query)
+	if err != nil {
+		p.log.Warn(err)
+	}
+	return trace
+}
+
+func (p *prober) getTraceForFinishedEvent() []byte {
+	query := fmt.Sprintf("cloudevents.type=%s and target=%s",
+		event.FinishedType, fmt.Sprintf(forwarderTargetFmt, p.client.Namespace))
```

Comment:
Do these queries work across source, broker, channel implementations?

Reply:
They should work because these Tags are added by Wathola sender when generating the events: [link](https://github.com/knative/eventing/blob/main/test/upgrade/prober/wathola/sender/services.go#L209). 
They're basically added at the beginning of the Trace and the queries return all traces that have these tags.
Right now it should work anywhere where the http sender is used because it properly passes the Context (propagates the trace). I still want to add it to the KafkaSender [here](https://github.com/knative-sandbox/eventing-kafka/blob/release-1.3/test/upgrade/continual/kafkasender.go#L67) (I have the code locally but waiting for the knative/eventing changes to propagate to that repo)

Reply:
@mgencur keep in mind that wathola sender has retries implemented (see [here](https://github.com/knative/eventing/blob/02663e2caa08c0ae96030caf348f1023d1d400b3/test/upgrade/prober/wathola/sender/services.go#L84-L111)). That means we could have multiple spans for a single step event, and we should report all of them.

Reply:
@cardil The query should return all traces for step/finished events with the given step/type as long as the event was not delivered in the end. For a missed event there might by multiple traces included in the report if the sender was trying to re-deliver them and it failed. And the Zipkin UI should be able to display them all.
I am not saving traces for duplicated events (as long as they were delivered) so I'm sending another commit that should fix it. Testing it now...